engine/src/main/java/org/kigalisim/engine/SingleThreadEngine.java

```diff
@@ -456,39 +762,342 @@ private UnitConverter createUnitConverterWithTotal(String stream) {
+      BigDecimal manufactureInitialChargeVal = manufactureInitialCharge.getValue();
+      BigDecimal importInitialChargeVal = unitConverter
+          .convert(importInitialCharge, manufactureInitialCharge.getUnits()).getValue();
+      BigDecimal totalInitialChargeVal = manufactureInitialChargeVal.add(importInitialChargeVal);
+      
+      if (totalInitialChargeVal.compareTo(BigDecimal.ZERO) == 0) {
+        percentManufacture = BigDecimal.ONE;
+        percentImport = BigDecimal.ZERO;
+      } else {
+        percentManufacture = manufactureInitialChargeVal.divide(totalInitialChargeVal, 10, java.math.RoundingMode.HALF_UP);
+        percentImport = importInitialChargeVal.divide(totalInitialChargeVal, 10, java.math.RoundingMode.HALF_UP);
+      }
+    } else {
+      percentManufacture = totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0 
+          ? BigDecimal.ZERO : manufactureSalesKg.divide(totalNonRecycleKg, 10, java.math.RoundingMode.HALF_UP);
+      percentImport = totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0 
+          ? BigDecimal.ZERO : importSalesKg.divide(totalNonRecycleKg, 10, java.math.RoundingMode.HALF_UP);
+    }
+
+    // Recycle
+    EngineNumber newRecycleValue = new EngineNumber(recycledDisplacedKg, "kg");
+    this.streamKeeper.setStream(application, substance, "recycle", newRecycleValue);
+
+    // New values
+    BigDecimal requiredKgUnbound = kgForRecharge.add(kgForNew);
+    BigDecimal requiredKg = requiredKgUnbound.compareTo(BigDecimal.ZERO) < 0 ? BigDecimal.ZERO : requiredKgUnbound;
+    BigDecimal newManufactureKg = percentManufacture.multiply(requiredKg);
+    BigDecimal newImportKg = percentImport.multiply(requiredKg);
+    EngineNumber newManufacture = new EngineNumber(newManufactureKg, "kg");
+    EngineNumber newImport = new EngineNumber(newImportKg, "kg");
+    this.streamKeeper.setStream(application, substance, "manufacture", newManufacture);
+    this.streamKeeper.setStream(application, substance, "import", newImport);
   }
 
   /**
-   * Placeholder for retirement recalculation.
+   * Recalculates retirement values based on current state.
    *
    * @param scope The scope to recalculate for
    */
   private void recalcRetire(Scope scope) {
-    // TODO: Implement recalc logic
+    // Setup
+    OverridingConverterStateGetter stateGetter =
+        new OverridingConverterStateGetter(this.stateGetter);
+    UnitConverter unitConverter = new UnitConverter(stateGetter);
+    Scope scopeEffective = scope != null ? scope : this.scope;
+    String application = scopeEffective.getApplication();
+    String substance = scopeEffective.getSubstance();
+
+    // Check allowed
+    if (application == null || substance == null) {
+      throw new RuntimeException(
+          "Tried recalculating population change without application and substance.");
+    }
+
+    // Calculate change
+    EngineNumber currentPriorRaw = this.streamKeeper.getStream(application, substance, "priorEquipment");
```

Comment:
Please do not use the `this.` prefix if possible here and elsewhere inside of `SingleThreadEngine`.